src/components/pinnable.js

```diff
@@ -0,0 +1,90 @@
+AFRAME.registerComponent("pinnable", {
+  schema: {
+    pinned: { default: false }
+  },
+
+  init() {
+    this._apply = this._apply.bind(this);
+    this._fireEvents = this._fireEvents.bind(this);
+    this._firePinnedEvents = this._firePinnedEvents.bind(this);
+    this._fireUnpinnedEvents = this._fireUnpinnedEvents.bind(this);
+    this._allowApplyOnceComponentsReady = this._allowApplyOnceComponentsReady.bind(this);
+    this._allowApply = false;
+
+    this.el.sceneEl.addEventListener("stateadded", this._apply);
+    this.el.sceneEl.addEventListener("stateremoved", this._apply);
+
+    // Fire pinned events when we drag and drop or scale in freeze mode,
+    // so transform gets updated.
+    this.el.addEventListener("grab-end", this._firePinnedEvents);
+
+    // Fire pinned events when page changes so we can persist the page.
+    this.el.addEventListener("pager-page-changed", this._firePinnedEvents);
+
+    // Hack: need to wait for the initial grabbable and stretchable components
+    // to show up from the template before applying.
+    this.el.addEventListener("componentinitialized", this._allowApplyOnceComponentsReady);
+    this._allowApplyOnceComponentsReady();
+  },
+
+  remove() {
+    this.el.sceneEl.removeEventListener("stateadded", this._apply);
+    this.el.sceneEl.removeEventListener("stateremoved", this._apply);
+    this.el.removeEventListener("componentinitialized", this._allowApplyOnceComponentsReady);
+  },
+
+  update() {
+    this._apply();
+    this._fireEvents();
+  },
+
+  _fireEvents() {
+    this._firePinnedEvents();
+    this._fireUnpinnedEvents();
+  },
+
+  _firePinnedEvents() {
+    if (this.data.pinned) {
+      this.el.emit("pinned");
+      this.el.sceneEl.emit("object_pinned", { el: this.el });
+    }
+  },
+
+  _fireUnpinnedEvents() {
+    if (!this.data.pinned) {
+      this.el.emit("unpinned");
+      this.el.sceneEl.emit("object_unpinned", { el: this.el });
+    }
+  },
+
+  _allowApplyOnceComponentsReady() {
+    if (!this._allowApply && this.el.components.grabbable && this.el.components.stretchable) {
+      if (this.el.components.grabbable.data.maxGrabbers !== 0) {
+        this.prevMaxGrabbers = this.el.components.grabbable.data.maxGrabbers;
```

Comment:
Seems like you would want to save this at the point you are about to overwrite it instead of just at load time, otherwise any changes that happened after the fact to maxGrabbers would be clobbered.

Reply:
most of the hairy bits of this have been GCed from my brain at this point, but I can assure you there were several revisions of this that went down that road :) I can take another go at it. A-Frame lifecycle fun